### PR TITLE
RNN DiagGGN extension

### PR DESCRIPTION
**Pull request by schaefertim**
Adds the DiagGGN extension for RNN module.
Refactor the DiagGGN extension.
Also adds second order support for Permute and ReduceTuple.

### CODE REVIEW

**Comment by schaefertim** [comment]
I'm a little bit unsatisfied with the current commit history because it is simply the complete history of my `rnn-develop` branch.
The actual changes are only the last 5 commits (the commit [reset to rnn branch](https://github.com/f-dangel/backpack/pull/159/commits/a4929d89764fbe9aa3c77721946030eff63e4302) is actually even with the `rnn` branch. 
Should I open a new pull request containing only the last 5 commits (although I don't know how to do that yet, so I would have to find that out)?

**Comment by f-dangel** [comment]
I will squash the commits, so it should not be a problem

File: fully_documented.txt [hunk 1]
```diff
@@ -4,6 +4,9 @@ backpack/core/derivatives/basederivatives.py
 backpack/core/derivatives/rnn.py
 backpack/core/derivatives/shape_check.py
 backpack/core/derivatives/__init__.py
+backpack/core/derivatives/permute.py
+
+backpack/extensions/backprop_extension.py
```
Comment (f-dangel) [inline fully_documented.txt:9]: Move to line 25 to keep directories grouped

File: test/extensions/secondorder/diag_ggn/test_batch_diag_ggn.py [hunk 2]
```diff
@@ -54,8 +56,9 @@ def test_diag_ggn_mc_batch_light(problem):
 @pytest.mark.montecarlo
 @pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
 def test_diag_ggn_mc_batch(problem):
-    """Test the MC approximation of individual diagonal of Gauss-Newton
-       with more samples (slow version)
+    """Test the MC approximation of individual diagonal.
+
+    of Gauss-Newton with more samples (slow version)
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_batch_diag_ggn.py:61]: add 'generalized'

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 3]
```diff
@@ -11,11 +12,12 @@
 
 
 @pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
-def test_diag_ggn(problem):
-    """Test the diagonal of Gauss-Newton
+def test_diag_ggn(problem, request):
+    """Test the diagonal of Gauss-Newton.
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:16]: add 'generalized'

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 4]
```diff
@@ -11,11 +12,12 @@
...
 
     Args:
         problem (ExtensionsTestProblem): Problem for extension test.
+        request: contains information about test
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:20]: remove

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 5]
```diff
@@ -33,8 +35,9 @@ def test_diag_ggn(problem):
 
 @pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
 def test_diag_ggn_mc_light(problem):
-    """Test the MC approximation of Diagonal of Gauss-Newton
-        with few mc_samples (light version)
+    """Test the MC approximation of Diagonal of Gauss-Newton.
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:38]: add 'generalized'

File: test/extensions/secondorder/diag_ggn/test_diag_ggn.py [hunk 6]
```diff
@@ -54,8 +57,9 @@ def test_diag_ggn_mc_light(problem):
 @pytest.mark.montecarlo
 @pytest.mark.parametrize("problem", PROBLEMS, ids=IDS)
 def test_diag_ggn_mc(problem):
-    """Test the MC approximation of Diagonal of Gauss-Newton
-       with more samples (slow version)
+    """Test the MC approximation of Diagonal of Gauss-Newton.
```
Comment (f-dangel) [inline test/extensions/secondorder/diag_ggn/test_diag_ggn.py:60]: add 'generalized'

File: backpack/extensions/secondorder/diag_ggn/diag_ggn_base.py [hunk 7]
```diff
@@ -1,6 +1,77 @@
...
+                diagonal
+            """
+            JS: Tensor = getattr(self.derivatives, f"{param}_jac_t_mat_prod")(
+                module, grad_inp, grad_out, backproped, sum_batch=False
+            )
+            return (
+                (JS ** 2).sum(axis=0).sum(axis=0)
+                if sum_batch
+                else (JS ** 2).sum(axis=0)
+            )
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/diag_ggn_base.py:75]:
```python
axis = [0,1] if sum_batch else 0
return (JS **2).sum(axis=axis)
```

File: backpack/extensions/secondorder/diag_ggn/rnn.py [hunk 8]
```diff
@@ -0,0 +1,27 @@
+"""Module implementing GGN."""
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/rnn.py:1]: Missing specification for `RNN`

File: backpack/extensions/secondorder/diag_ggn/rnn.py [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+from backpack.core.derivatives.rnn import RNNDerivatives
+from backpack.extensions.secondorder.diag_ggn.diag_ggn_base import DiagGGNBaseModule
+
+
+class DiagGGNRNN(DiagGGNBaseModule):
+    """Calculating GGN derivative."""
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/rnn.py:7]: Strictly speaking, the GGN is not a derivative but a matrix. Replace with 'diagonal'

File: backpack/extensions/secondorder/diag_ggn/rnn.py [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+        """Initialize."""
+        super().__init__(
+            derivatives=RNNDerivatives(),
+            params=["bias_ih_l0", "bias_hh_l0", "weight_ih_l0", "weight_hh_l0"],
+            sum_batch=True,
+        )
+
+
+class BatchDiagGGNRNN(DiagGGNBaseModule):
+    """Calculating GGN derivative."""
```
Comment (f-dangel) [inline backpack/extensions/secondorder/diag_ggn/rnn.py:19]: Strictly speaking, the GGN is not a derivative but a matrix. Replace with 'per-sample diagonal'

File: fully_documented.txt (hunk 1 above)
Comment (schaefertim) [reply fully_documented.txt:9]: Can you please double-check whether it's really better?
I feel like grouping all of `backpack/extensions/` together would go too far. I prefer having the sub-folders of it separate.

File: backpack/core/derivatives/permute.py [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+        self, module: Permute, g_inp: Tuple[Tensor], g_out: Tuple[Tensor], mat: Tensor
+    ) -> Tensor:
+        return mat.permute(
+            [0] + [element + 1 for element in argsort(Tensor(module.dims))]
+        )
+
+    def _jac_mat_prod(
+        self, module: Permute, g_inp: Tuple[Tensor], g_out: Tuple[Tensor], mat: Tensor
+    ) -> Tensor:
+        return mat.permute([0] + [element + 1 for element in module.dims])
```
[the remark this answers is not in the document]
Comment (schaefertim) [reply backpack/core/derivatives/permute.py:23]: Tests added.

**Pull request merged**